File: MagicHax/android/app/src/main/java/me/nathanp/magiccreatures/view/CardAdapter.java
package me.nathanp.magiccreatures.view;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import me.nathanp.magiccreatures.R;

public class CardAdapter<T extends CardAdapter.CardInfo> extends RecyclerView.Adapter<CardAdapter<T>.CardHolder> {

    public interface onCardSelectedListener<O> {
        void onSelected(O thing);
    }

    public interface CardInfo {
        int getDrawableId();
        String getName();
        String getDescription();
    }

    private List<T> things = new ArrayList<>();
    private onCardSelectedListener<T> listener;

    public CardAdapter(onCardSelectedListener<T> listener) {
        this.listener = listener;
    }

    @NonNull
    @Override
    public CardHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.card_item, parent, false);
        return new CardHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull CardHolder holder, int position) {
        T thing = things.get(position);
        holder.bind(thing);
    }

    @Override
    public int getItemCount() {
        return things.size();
    }

    public void setThings(List<T> things) {
        this.things = things;
        notifyDataSetChanged();
    }

    class CardHolder extends RecyclerView.ViewHolder {
        T myThing;
        private ImageView imageView;
        private TextView textViewName;
        private TextView textViewStats;

        CardHolder(View view) {
            super(view);
            view.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    listener.onSelected(myThing);
                }
            });
            imageView = view.findViewById(R.id.card_item_image);
            textViewName = view.findViewById(R.id.card_item_name);
            textViewStats = view.findViewById(R.id.creature_item_stats);
        }

        void bind(T thing) {
            myThing = thing;
            itemView.setTag(thing);
            Glide.with(imageView.getContext())
                    .load(thing.getDrawableId())
                    .into(imageView);
            textViewName.setText(thing.getName());
            textViewStats.setText(thing.getDescription());
        }
    }
}

File: MagicHax/android/app/src/main/java/me/nathanp/magiccreatures/view/CardAdapter.kt
package me.nathanp.magiccreatures.view

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import me.nathanp.magiccreatures.databinding.CardItemBinding
import me.nathanp.magiccreatures.view.CardAdapter.CardInfo
import java.util.*

abstract class CardAdapter<T : CardInfo> : RecyclerView.Adapter<CardAdapter<T>.CardHolder>() {
    abstract fun onSelected(pos: Int, item: T)
    interface CardInfo {
        val drawableId: Int
        val name: String
        val description: String
        val isCheckable: Boolean
        var checked: Boolean
    }

    var things: List<T> = ArrayList()
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CardHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = CardItemBinding.inflate(inflater, parent, false)
        return CardHolder(binding)
    }

    override fun onBindViewHolder(holder: CardHolder, position: Int) {
        holder.bind(things[position])
    }

    override fun getItemCount() = things.size

    inner class CardHolder(var binding: CardItemBinding) : RecyclerView.ViewHolder(binding.root) {
        private lateinit var thing: T

        init {
            binding.root.setOnClickListener {
                binding.checkBox.isChecked = !binding.checkBox.isChecked
                onSelected(adapterPosition, thing)
            }
        }

        fun bind(thing: T) {
            this.thing = thing
            binding.cardinfo = thing
            binding.executePendingBindings()
        }
    }
}